File: frontend/src/lib/vendors.ts
// ---------------------------------------------------------------------------
// Vendor data for the /admin/vendors table
// ---------------------------------------------------------------------------
// Mirrors lib/customers.ts. Fetches from GET /admin/get_vendor_details
// (not yet implemented on the backend — see backend/app/models/vendor_details.py
// and vendor_poc_details.py for the underlying collections), which is expected
// to return every vendor as a flat list. `id` (VendorDetails' Beanie primary
// key) is the unique key used for both the table's React key and matching a
// row back to a vendor for editing — unlike customers, vendors have no email
// to key off of.
//
// `qrCode` is the vendor's standard UPI deep-link string (e.g.
// "upi://pay?pa=vendor@upibank&pn=Vendor&..."), decoded from an uploaded QR
// code image via POST /admin/convert_vendor_qr (see convertVendorQr below).
import { apiFetch } from "@/lib/api";

export type VendorPoc = {
  name: string;
  phone: string;
};

export type Vendor = {
  id: number;
  registeredName: string;
  gst: string;
  address: string;
  description: string;
  qrCode: string;
  isDeleted: boolean;
  pocs: VendorPoc[];
};

// Shape expected from the backend's future VendorDetailItem schema —
// mirrors CustomerDetailItem's contact_name/contact_phone array pairing.
type VendorDetailItem = {
  id: number;
  registered_name: string;
  gst: string;
  address: string;
  description: string;
  qr_code: string;
  is_deleted: boolean;
  contact_name: string[];
  contact_phone: string[];
};

export type VendorOption = {
  id: number;
  name: string;
  // "" if the vendor has no GST number on file — see product-form-modal.tsx,
  // which filters these down to GST-only for its vendor picker.
  gst: string;
};

type VendorListItem = {
  vendor_id: number;
  vendor_name: string;
  gst: string;
};

// Lightweight id+name list for vendor-picker dropdowns (product and purchase
// order popups) — GET /admin/get_vendors_list only returns active vendors,
// unlike fetchVendors's full get_vendor_details.
export async function fetchVendorsList(): Promise<VendorOption[]> {
  const response = await apiFetch("/admin/get_vendors_list");
  if (!response.ok) {
    throw new Error("Failed to load vendors");
  }

  const items: VendorListItem[] = await response.json();
  return items.map((item) => ({ id: item.vendor_id, name: item.vendor_name, gst: item.gst }));
}

export async function fetchVendors(): Promise<Vendor[]> {
  const response = await apiFetch("/admin/get_vendor_details");
  if (!response.ok) {
    throw new Error("Failed to load vendors");
  }

  const items: VendorDetailItem[] = await response.json();
  return items.map((item) => ({
    id: item.id,
    registeredName: item.registered_name,
    gst: item.gst,
    address: item.address,
    description: item.description,
    qrCode: item.qr_code,
    isDeleted: item.is_deleted,
    pocs: item.contact_name.map((name, index) => ({
      name,
      phone: item.contact_phone[index] ?? "",
    })),
  }));
}

export type ConvertedQr = {
  qrCode: string;
  // "QR code added" / "QR code updated" — depends on whether `vendorId`
  // already had a qr_code set, so it's only meaningful when a vendorId was
  // passed (i.e. editing an existing vendor).
  message: string;
};

// Uploads a vendor's QR code image and returns its decoded standard UPI
// string. Rejects (with the backend's detail message) if the image has no
// QR code, or the QR code isn't a "upi://pay?..." payment link. `vendorId`
// is only passed in "edit" mode — see convert_vendor_qr in
// backend/app/api/routes/vendors.py for how it's used to pick the message.
export async function convertVendorQr(file: File, vendorId?: number): Promise<ConvertedQr> {
  const formData = new FormData();
  formData.append("file", file);
  if (vendorId !== undefined) {
    formData.append("vendor_id", String(vendorId));
  }

  const response = await apiFetch("/admin/convert_vendor_qr", {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const detail = await response.json().catch(() => null);
    throw new Error(detail?.detail ?? "Failed to read QR code");
  }

  const { qr_code: qrCode, message }: { qr_code: string; message: string } = await response.json();
  return { qrCode, message };
}

File: backend/app/api/routes/vendors.py
# Vendors module: endpoints for managing vendor accounts and their points of
# contact, restricted to admins (bypassed entirely when settings.auth_enabled
# is False, matching require_admin in routes/admin.py).
from beanie.operators import In
from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status

from app.api.routes.admin import party_state_or_400, require_admin
from app.models import User, VendorDetails, VendorIdCounter, VendorPocDetails, VendorPocIdCounter
from app.schemas.vendors import (
    AddVendorDetailsRequest,
    AddVendorDetailsResponse,
    ConvertVendorQrResponse,
    UpdateVendorDetailsRequest,
    UpdateVendorDetailsResponse,
    VendorDetailItem,
    VendorListItem,
)
from app.services.counters import get_next_id
from app.services.qr import InvalidQrCodeError, decode_upi_qr

router = APIRouter(prefix="/admin", tags=["vendors"])


@router.post("/add_vendor_details", response_model=AddVendorDetailsResponse)
async def add_vendor_details(
    payload: AddVendorDetailsRequest,
    _: User | None = Depends(require_admin),
) -> AddVendorDetailsResponse:
    state_code, state_name = party_state_or_400(payload.state_code, payload.gst)

    vendor_id = await get_next_id(VendorIdCounter, "next_vendor_id", VendorDetails)
    vendor = VendorDetails(
        id=vendor_id,
        registered_name=payload.registered_name,
        gst=payload.gst,
        state_code=state_code,
        state_name=state_name,
        address=payload.address,
        description=payload.description,
        vendor_type=payload.vendor_type,
        is_deleted=payload.is_deleted,
    )
    await vendor.insert()

    for contact_name, contact_phone in zip(payload.contact_name, payload.contact_phone):
        poc_id = await get_next_id(VendorPocIdCounter, "next_vendor_poc_id", VendorPocDetails)
        poc = VendorPocDetails(id=poc_id, vendor_id=vendor_id, contact_name=contact_name, contact_phone=contact_phone)
        await poc.insert()

    return AddVendorDetailsResponse(message="vendor successfully added")


@router.get("/get_vendors_list", response_model=list[VendorListItem])
async def get_vendors_list(
    _: User | None = Depends(require_admin),
) -> list[VendorListItem]:
    # Lightweight id+name list for vendor-picker dropdowns (e.g. the product
    # and purchase order popups) — unlike get_vendor_details, this excludes
    # soft-deleted vendors and skips the address/POC lookups those popups
    # don't need. gst rides along (rather than being skipped like
    # address/POC) so the product form's picker can filter down to
    # GST-invoiceable vendors only — see product-form-modal.tsx.
    vendors = await VendorDetails.find(VendorDetails.is_deleted == False).to_list()
    return [
        VendorListItem(
            vendor_id=vendor.id,
            vendor_name=vendor.registered_name,
            gst=vendor.gst,
            state_code=vendor.state_code,
            state_name=vendor.state_name,
        )
        for vendor in vendors
    ]


@router.get("/get_vendor_details", response_model=list[VendorDetailItem])
async def get_vendor_details(
    _: User | None = Depends(require_admin),
) -> list[VendorDetailItem]:
    vendors = await VendorDetails.find_all().to_list()
    if not vendors:
        return []

    vendor_ids = [vendor.id for vendor in vendors]
    pocs = await VendorPocDetails.find(In(VendorPocDetails.vendor_id, vendor_ids)).to_list()
    pocs_by_vendor_id: dict[int, list[VendorPocDetails]] = {}
    for poc in pocs:
        pocs_by_vendor_id.setdefault(poc.vendor_id, []).append(poc)

    response = []
    for vendor in vendors:
        vendor_pocs = pocs_by_vendor_id.get(vendor.id, [])
        response.append(
            VendorDetailItem(
                id=vendor.id,
                registered_name=vendor.registered_name,
                gst=vendor.gst,
                state_code=vendor.state_code,
                state_name=vendor.state_name,
                address=vendor.address,
                description=vendor.description,
                qr_code=vendor.qr_code,
                vendor_type=vendor.vendor_type,
                is_deleted=vendor.is_deleted,
                contact_name=[poc.contact_name for poc in vendor_pocs],
                contact_phone=[poc.contact_phone for poc in vendor_pocs],
            )
        )

    return response


@router.post("/update_vendor_details", response_model=UpdateVendorDetailsResponse)
async def update_vendor_details(
    payload: UpdateVendorDetailsRequest,
    _: User | None = Depends(require_admin),
) -> UpdateVendorDetailsResponse:
    vendor = await VendorDetails.get(payload.id)
    if vendor is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="vendor not found")

    vendor.registered_name = payload.registered_name
    vendor.gst = payload.gst
    vendor.state_code, vendor.state_name = party_state_or_400(payload.state_code, payload.gst)
    vendor.address = payload.address
    vendor.qr_code = payload.qr_code
    vendor.description = payload.description
    vendor.vendor_type = payload.vendor_type
    vendor.is_deleted = payload.is_deleted
    await vendor.save()

    await VendorPocDetails.find(VendorPocDetails.vendor_id == vendor.id).delete()
    for contact_name, contact_phone in zip(payload.contact_name, payload.contact_phone):
        poc_id = await get_next_id(VendorPocIdCounter, "next_vendor_poc_id", VendorPocDetails)
        poc = VendorPocDetails(id=poc_id, vendor_id=vendor.id, contact_name=contact_name, contact_phone=contact_phone)
        await poc.insert()

    return UpdateVendorDetailsResponse(message="vendor updated successfully")


@router.post("/convert_vendor_qr", response_model=ConvertVendorQrResponse)
async def convert_vendor_qr(
    file: UploadFile = File(...),
    # Only present in "edit" mode (see components/admin/vendor-form-modal.tsx)
    # — lets the message below distinguish a vendor's first QR from a
    # replacement of one they already had. Omitted entirely for a
    # brand-new vendor, which is always "added".
    vendor_id: int | None = Form(default=None),
    _: User | None = Depends(require_admin),
) -> ConvertVendorQrResponse:
    image_bytes = await file.read()
    try:
        qr_code = decode_upi_qr(image_bytes)
    except InvalidQrCodeError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))

    had_existing_qr = False
    if vendor_id is not None:
        vendor = await VendorDetails.get(vendor_id)
        had_existing_qr = bool(vendor and vendor.qr_code)

    message = "QR code updated" if had_existing_qr else "QR code added"
    return ConvertVendorQrResponse(qr_code=qr_code, message=message)
